# simRiboSeq

| Module             	| Function                                                                      	| Input                                                                                                                                                                                                	| Output                                                                              	|
|--------------------	|-------------------------------------------------------------------------------	|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------	|-------------------------------------------------------------------------------------	|
| sim_transcriptome.R 	| generate_transcript()                                                                 	| - Number of codons in transcript<br>- Genomic codon distributions                                                                                                                                    	| Individual transcript codon sequence*                                               	|
|                    	| simulate_transcriptome()  * plus padding (6 codons for 5’ end, 4 codons for 3’ end) 	| - Transcript lengths (in codons) for all transcripts<br>- Genomic codon distributions                                                                                                                	| .fa file for transcriptome*                                                         	|
| simulate_ribosome_distribution.R      	| simulate_pi()                                                                       	| - Transcriptome* .fa file<br>- Transcript lengths from model experiment<br>- Transcript abundances from model experiment<br>- Model class for abundance ~ length (loess / linear regression)         	| Vector of probabilities of ribosome mapping to individual transcript                	|
|                    	| simulate_rho()                                                                      	| - Transcriptome* .fa file<br>- Number of codons padding cds<br>- Translational efficiency of codon averaged over genome                                                                              	| Vector of probabilities of ribosome mapping to individual codon within a transcript 	|
| simulate_footprints.R    	| digest() - digest_transcript()                                                	| - Transcriptome* .fa file<br>- Ribosome counts (rawProfiles.txt)<br>- Probabilities for 5’ digest lengths<br>- Probabilities for 3’ digest lengths<br>- Minimum read length<br>- Maximum read length 	| Footprint sequences                                                                 	|
|                    	| get_bias_region()                                                               	| - Footprint sequences<br>- Length of bias region                                                                                                                                                     	| Nucleotide sequence for bias region                                                 	|
|                    	| ligate()                                                                      	| - Footprint sequences<br>- Probabilities of successful ligation for bias regions                                                                                                                     	| Footprint sequences                                                                 	|
|                    	| circularize()                                                                 	| - Footprint sequences<br>- Probabilities of successful circularization for bias regions                                                                                                              	| Footprint sequences                                                                 	|
|                    	| simulate_footprints()                                                               	| [ wrapper function ]<br>- sequencing sampling loss                                                                                                                                                   	| .fa file for footprint sequences                                                    	|
| helper.R           	| read_fasta()                                                                  	|                                                                                                                                                                                                      	|                                                                                     	|
|                    	| read_raw_profiles()                                                             	|                                                                                                                                                                                                      	|                                                                                     	|
